serial/pet/driver.s

```diff
@@ -0,0 +1,730 @@
+                ;      110  300  600 1200 2400 4800 9600 (Baud)
+KbdPollIntTbl:     .byte 2,   5,  10,  20,  40,  80, 160
+      ; Poll freq Hz    55   60   60   60   60   60   60
+                ; If KbdPollIntrvl value is below 12 we need to use the all at once keyboard scan
+
+                ; Timer 2 isn't freerunning so we have to subtract the cycles till we reset
+                ; it from the rate (- $5D)
+Tim2BaudLo:     .byte $26, $A8, $26, $e4, $44, $73, $0B
+Tim2BaudHi:     .byte $23, $0C, $06, $02, $01, $00, $00
+
+                ; A log2 table allows quick decoding by giving the index of the highest bit set
+                ; Remembering that modifers need to be checked seperatly
+                ; This table is inverted due to our key tables being backwards
+                ; 7 - LOG_2(x)
+Log2Tbl:        .byte 255,7,6,6,5,5,5,5,4,4,4,4,4,4,4,4
+                .byte   3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3
+                .byte   2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2
+                .byte   2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2
+
+;----------------------------------------------------------------------
+; $00 = shift
+; $EF = non-existant
+; $FF = REV  (ctrl key)
+; $F0 = HOME (MENU key)
+;
+; $F1 = UP
+; $F2 = DOWN
+; $F3 = RIGHT
+; $F4 = LEFT
+;
+.if BSNSS_KBD
+
+                ; Matrix for Business Keyboards
+KbdMatrix:      .byte '@', $0E, $F3, '8', '-', '8', '5', '2'    ; $8E = BothShift+2, $1D = CursRight
+                .byte '9', $EF, '^', '7', '0', '7', '4', '1'
+                .byte '5', '\', 'k', ';', 'h', 'f', 's', $1B    ; $9B = ESC
+                .byte '6', '[', 'l', $0D, 'j', 'g', 'd', 'a'
+                .byte $08, 'p', 'i', '@', 'y', 'r', 'w', $09    ; $C0 = nonshiftable ., $FF= nonshift DEL
+                .byte '4', ']', 'o', $F2, 'u', 't', 'e', 'q'  ; $91 = CursUP.
+                .byte '3', $00, $19, '.', '.', 'b', 'c', $00    ; $AE-> KP.
+                .byte '2', $04, $0F, '0', $2C, 'n', 'v', 'z'    ; Repeat->^D, $0F = Z+A+L??
+                .byte '1', '/', $15, $F0, 'm', ' ', 'x', $FF    ; $15 - RVS + A + L??, B1 = KP1
+                .byte $16, $EF, ':', $03, '9', '6', '3', $08    ; $88 Left Arrow to BS?, ^V=TAB+<-+DEL
+
+                ; Keymasks to remove modifers from the scan results
+                ; There are backward of the table above! Above goes from 9->0, these are 0->9
+KeyMask:        .byte $FF, $BF, $FF, $FF, $FF, $FF, $BE, $FF, $FE, $BF
+                ; Which bits indicate shift keys
+ShiftMask:      .byte $00, $00, $00, $00, $00, $00, $41, $00, $00, $00
+                ; Which bits indicate ctrl keys
+CtrlMask:       .byte $00, $00, $00, $00, $00, $00, $00, $00, $01, $00
+
+                ; Keyboard matrix with shift pressed, needed for consistent shifts
+                ; Matrix for Business Keyboards
+KbdMatrixShift: .byte '>', $0E, $F4, '8', '=', '(', '%', '"'    ;" ;$8E = BothShift+2, $9D = CursRight
```

Comment:
I'm just going to assume this is right :-)

Reply:
As did I. :) We can't really test this without a PET with a business keyboard, I think.